stdlib/Makefile

```diff
@@ -13,66 +13,236 @@
+  digest.cmo random.cmo hashtbl.cmo weak.cmo \
+  format.cmo uchar.cmo scanf.cmo callback.cmo \
+  camlinternalOO.cmo oo.cmo camlinternalMod.cmo \
+  genlex.cmo ephemeron.cmo \
+  filename.cmo complex.cmo \
+  arrayLabels.cmo listLabels.cmo bytesLabels.cmo \
+  stringLabels.cmo moreLabels.cmo stdLabels.cmo \
+  spacetime.cmo
+
+.PHONY: all
+all: stdlib.cma std_exit.cmo camlheader target_camlheader camlheader_ur
+
+ifeq "$(RUNTIMED)" "runtimed"
+all: camlheaderd
+endif
+
+ifeq "$(RUNTIMEI)" "true"
+all: camlheaderi
+endif
+
+.PHONY: allopt
+allopt: stdlib.cmxa std_exit.cmx allopt-$(PROFILING)
+
+.PHONY: allopt-noprof
 allopt-noprof:
 
+.PHONY: allopt-prof
 allopt-prof: stdlib.p.cmxa std_exit.p.cmx
 	rm -f std_exit.p.cmi
 
+INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
+
+.PHONY: install
+install::
+	cp stdlib.cma std_exit.cmo *.cmi *.cmt *.cmti *.mli *.ml \
+	  camlheader_ur \
+	  "$(INSTALL_LIBDIR)"
+	cp target_camlheader "$(INSTALL_LIBDIR)/camlheader"
+
+ifeq "$(RUNTIMED)" "runtimed"
+install::
+	cp target_camlheaderd $(INSTALL_LIBDIR)
+endif
+
+ifeq "$(RUNTIMEI)" "true"
+install::
+	cp target_camlheaderi $(INSTALL_LIBDIR)
+endif
+
+.PHONY: installopt
 installopt: installopt-default installopt-$(PROFILING)
 
+.PHONY: installopt-default
 installopt-default:
-	cp stdlib.cmxa stdlib.a std_exit.o *.cmx "$(INSTALL_LIBDIR)"
-	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.a
+	cp stdlib.cmxa stdlib.$(A) std_exit.$(O) *.cmx "$(INSTALL_LIBDIR)"
+	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.$(A)
 
+.PHONY: installopt-noprof
 installopt-noprof:
-	rm -f "$(INSTALL_LIBDIR)/stdlib.p.cmxa"; \
-	  ln -s stdlib.cmxa "$(INSTALL_LIBDIR)/stdlib.p.cmxa"
-	rm -f "$(INSTALL_LIBDIR)/stdlib.p.a"; \
-	  ln -s stdlib.a "$(INSTALL_LIBDIR)/stdlib.p.a"
-	rm -f "$(INSTALL_LIBDIR)/std_exit.p.cmx"; \
-	  ln -s std_exit.cmx "$(INSTALL_LIBDIR)/std_exit.p.cmx"
-	rm -f "$(INSTALL_LIBDIR)/std_exit.p.o"; \
-	  ln -s std_exit.o "$(INSTALL_LIBDIR)/std_exit.p.o"
+	ln -sf stdlib.cmxa "$(INSTALL_LIBDIR)/stdlib.p.cmxa"
+	ln -sf stdlib.$(A) "$(INSTALL_LIBDIR)/stdlib.p.$(A)"
+	ln -sf std_exit.cmx "$(INSTALL_LIBDIR)/std_exit.p.cmx"
+	ln -sf std_exit.$(O) "$(INSTALL_LIBDIR)/std_exit.p.$(O)"
 
+.PHONY: installopt-prof
 installopt-prof:
-	cp stdlib.p.cmxa stdlib.p.a std_exit.p.cmx std_exit.p.o \
+	cp stdlib.p.cmxa stdlib.p.$(A) std_exit.p.cmx std_exit.p.$(O) \
 	  "$(INSTALL_LIBDIR)"
-	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.p.a
+	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.p.$(A)
+
+ifeq "$(UNIX_OR_WIN32)" "unix"
+HEADERPROGRAM = header.c
+else # Windows
+HEADERPROGRAM = headernt.c
+endif
+
+CAMLHEADERS =\
+  camlheader target_camlheader camlheader_ur \
+  camlheaderd target_camlheaderd \
+  camlheaderi target_camlheaderi
+
+ifeq "$(HASHBANGSCRIPTS)" "true"
+$(CAMLHEADERS): ../config/Makefile
+	for suff in '' d i; do \
+	  echo '#!$(BINDIR)/ocamlrun'$$suff > camlheader$$suff && \
+	  echo '#!$(TARGET_BINDIR)/ocamlrun'$$suff >target_camlheader$$suff; \
+	done && \
+	echo '#!' | tr -d '\012' > camlheader_ur;
```

Comment:
This is most probably for another patchset:

    printf "#!" > camlheader_ur

would be much cleaner. I'm not sure we can get into this branch on platforms with \r\n end-of-lines however.